File: tests/data-conversion/class-inherit/inherit.ts
class Prop {}

class Base {
  public prop: Prop;

  constructor(prop: Prop) {
    this.prop = prop;
  }

  method(arg?: Prop) {}
}

class Derived extends Base {
  public childProp?: Prop;

  constructor() {
    super(new Prop());
  }

  override method(arg?: Prop) {
    super.method(arg);
  }
}

class NotDerived {
}

function TestInheritance() {
  const base: Base = new Derived();
}

File: tests/data-conversion/class-inherit/inherit.cpp
#include "runtime/object.h"

namespace {

class Prop : public compilets::Object {
};

class Base : public compilets::Object {
 public:
  cppgc::Member<Prop> prop;

  Base(Prop* prop) {
    this->prop = prop;
  }

  virtual void method(Prop* arg) {}

  void Trace(cppgc::Visitor* visitor) const override {
    compilets::TraceMember(visitor, prop);
  }

  virtual ~Base() = default;
};

class Derived : public Base {
 public:
  cppgc::Member<Prop> childProp;

  Derived() : Base(compilets::MakeObject<Prop>()) {}

  void method(Prop* arg) override {
    Base::method(arg);
  }

  void Trace(cppgc::Visitor* visitor) const override {
    compilets::TraceMember(visitor, childProp);
    Base::Trace(visitor);
  }

  virtual ~Derived() = default;
};

class NotDerived : public compilets::Object {
};

void TestInheritance() {
  Base* base = compilets::MakeObject<Derived>();
}

}  // namespace
